url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/SSLSocketChannel2.java

```diff
@@ -0,0 +1,380 @@
+			socketChannel.write( wrap( emptybuffer ) );
+			if( writeEngineResult.getHandshakeStatus() == HandshakeStatus.FINISHED ) {
+				createBuffers( sslEngine.getSession() );
+				return;
+			}
+		}
+		assert ( sslEngine.getHandshakeStatus() != HandshakeStatus.NOT_HANDSHAKING );// this function could only leave NOT_HANDSHAKING after createBuffers was called unless #190 occurs which means that nio wrap/unwrap never return HandshakeStatus.FINISHED
+
+		bufferallocations = 1; // look at variable declaration why this line exists and #190. Without this line buffers would not be be recreated when #190 AND a rehandshake occur.
+	}
+	private synchronized ByteBuffer wrap( ByteBuffer b ) throws SSLException {
+		outCrypt.compact();
+		writeEngineResult = sslEngine.wrap( b, outCrypt );
+		outCrypt.flip();
+		return outCrypt;
+	}
+
+	/**
+	 * performs the unwrap operation by unwrapping from {@link #inCrypt} to {@link #inData}
+	 **/
+	private synchronized ByteBuffer unwrap() throws SSLException {
+		int rem;
+		do {
+			rem = inData.remaining();
+			readEngineResult = sslEngine.unwrap( inCrypt, inData );
+		} while ( readEngineResult.getStatus() == SSLEngineResult.Status.OK && ( rem != inData.remaining() || sslEngine.getHandshakeStatus() == HandshakeStatus.NEED_UNWRAP ) );
+		inData.flip();
+		return inData;
+	}
+
+	protected void consumeDelegatedTasks() {
+		Runnable task;
+		while ( ( task = sslEngine.getDelegatedTask() ) != null ) {
+			tasks.add( exec.submit( task ) );
+			// task.run();
+		}
+	}
+
+	protected void createBuffers( SSLSession session ) {
+		int appBufferMax = session.getApplicationBufferSize();
+		int netBufferMax = session.getPacketBufferSize();
+
+		if( inData == null ) {
+			inData = ByteBuffer.allocate( appBufferMax );
+			outCrypt = ByteBuffer.allocate( netBufferMax );
+			inCrypt = ByteBuffer.allocate( netBufferMax );
+		} else {
+			if( inData.capacity() != appBufferMax )
+				inData = ByteBuffer.allocate( appBufferMax );
+			if( outCrypt.capacity() != netBufferMax )
+				outCrypt = ByteBuffer.allocate( netBufferMax );
+			if( inCrypt.capacity() != netBufferMax )
+				inCrypt = ByteBuffer.allocate( netBufferMax );
+		}
+		inData.rewind();
+		inData.flip();
+		inCrypt.rewind();
+		inCrypt.flip();
+		outCrypt.rewind();
+		outCrypt.flip();
+		bufferallocations++;
+	}
+
+	public int write( ByteBuffer src ) throws IOException {
+		if( !isHandShakeComplete() ) {
+			processHandshake();
+			return 0;
+		}
+		// assert ( bufferallocations > 1 ); //see #190
+		if( bufferallocations <= 1 ) {
+			createBuffers( sslEngine.getSession() );
+		}
+		int num = socketChannel.write( wrap( src ) );
+		return num;
+
+	}
+
+	/**
+	 * Blocks when in blocking mode until at least one byte has been decoded.<br>
+	 * When not in blocking mode 0 may be returned.
+	 * 
+	 * @return the number of bytes read.
+	 **/
+	public int read( ByteBuffer dst ) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 229-240 into a separate method.